sdk/core/Azure.Core/tests/OperationInternalTests.cs

```diff
@@ -0,0 +1,361 @@
+            };
+            var operationInternal = testOperation.OperationInternal;
+
+            try
+            {
+                _ = async
+                    ? await operationInternal.UpdateStatusAsync(CancellationToken.None)
+                    : operationInternal.UpdateStatus(CancellationToken.None);
+            }
+            catch { }
+
+            testListener.AssertScopeException($"{nameof(TestOperation)}.UpdateStatus", (Exception scopeException) =>
+            {
+                Assert.AreEqual(originalException, scopeException);
+            });
+        }
+
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task UpdateStatusSetsFailedScopeWhenOperationThrows(bool async)
+        {
+            using var testListener = new ClientDiagnosticListener(DiagnosticNamespace);
+
+            var originalException = new StackOverflowException();
+            var testOperation = new TestOperation()
+            {
+                OnUpdateState = _ => throw originalException
+            };
+            var operationInternal = testOperation.OperationInternal;
+
+            try
+            {
+                _ = async
+                    ? await operationInternal.UpdateStatusAsync(CancellationToken.None)
+                    : operationInternal.UpdateStatus(CancellationToken.None);
+            }
+            catch { }
+
+            testListener.AssertScopeException($"{nameof(TestOperation)}.UpdateStatus", (Exception scopeException) =>
+            {
+                Assert.IsInstanceOf(typeof(RequestFailedException), scopeException);
+                Assert.AreEqual(originalException, scopeException.InnerException);
+            });
+        }
+
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task UpdateStatusPassesTheCancellationTokenToUpdateState(bool async)
+        {
+            using var tokenSource = new CancellationTokenSource();
+            var originalToken = tokenSource.Token;
+            CancellationToken passedToken = default;
+
+            var mockResponse = new MockResponse(200);
+            var testOperation = new TestOperation()
+            {
+                OnUpdateState = cancellationToken =>
+                {
+                    passedToken = cancellationToken;
+                    return OperationState<int>.Pending(mockResponse);
+                }
+            };
+            var operationInternal = testOperation.OperationInternal;
+
+            _ = async
+                ? await operationInternal.UpdateStatusAsync(originalToken)
+                : operationInternal.UpdateStatus(originalToken);
+
+            Assert.AreEqual(originalToken, passedToken);
+        }
+
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task WaitForCompletionCallsUntilOperationCompletes(bool useDefaultPollingInterval)
```

Comment:
Is there an ideal deterministic way to test how much it takes between each `UpdateState` call so we can measure the delay used? We could add tests for the polling interval selection algorithm (default, retry-after header, or `pollingInterval` parameter).

We could have a separate method for the delay in the `OperationInternal` class:
```cs
protected async virtual Delay(TimeSpan interval) => await Task.Delay(interval).ConfigureAwait(false);
```
And we could mock it with a child test class to check the `interval` used.

Not sure if there's a better way.

Reply:
>We could have a separate method for the delay in the OperationInternal class:
>protected async virtual Delay(TimeSpan interval) => await Task.Delay(interval).ConfigureAwait(false);

I like this one. We do it for the RetryPolicy as well.

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs#L140-L149

Reply:
Updated.